lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,517 @@
+///          toOp(*it);
+///      }
+///
+///  We can see that for each next operation in btor2, we will compute all
+///  the prerequisite operations before storing the result in our cache
+///===----------------------------------------------------------------------===//
+void Deserialize::toOp(Btor2Line *line) {
+  if (cache.find(line->id) != cache.end()) {
+    return;
+  }
+
+  Operation *res = nullptr;
+  std::vector<Btor2Line *> todo;
+  todo.push_back(line);
+  while (!todo.empty()) {
+    auto cur = todo.back();
+    uint32_t oldsize = todo.size();
+    for (uint32_t i = 0; i < cur->nargs; ++i) {
+      if (cur->args[i] > 0 
+      && !isValidChild(reachedLines.at(cur->args[i]))) {
+        continue;
+      }
+
+      if (cache.find(cur->args[i]) == cache.end()) {
+        if (cur->args[i] < 0) {
+          // if original operation is cached, negate it
+          if (cache.find(cur->args[i] * -1) != cache.end()) {
+            createNegateLine(cur->args[i], cache.at(cur->args[i] * -1)); 
+          } else {
+            todo.push_back(reachedLines.at(cur->args[i] * -1));
+          }
+        } else {
+          todo.push_back(reachedLines.at(cur->args[i]));
+        }
+      }
+    }
+    if (todo.size() != oldsize) {
+      continue;
+    }
+    if (!isValidChild(cur) 
+    || cache.find(cur->id) != cache.end()) {
+      todo.pop_back();
+      continue;
+    }
+    res = createMLIR(cur, cur->args);
+    assert(res && res->getNumResults() < 2 && res->getResult(0));
+    cache[cur->id] = res->getResult(0);
+    todo.pop_back();
+  }
+}
+
+OwningOpRef<FuncOp> Deserialize::buildInitFunction() {
+  // collect the return types for our init function
+  std::vector<Type> returnTypes(states.size(), nullptr);
+  for (uint32_t i = 0; i < states.size(); ++i) {
+    returnTypes[i] = builder.getIntegerType(states.at(i)->sort.bitvec.width);
+    assert(returnTypes[i]);
+  }
+
+  // create init function signature
+  OperationState state(unknownLoc, FuncOp::getOperationName());
+  FuncOp::build(builder, state, "init",
+                FunctionType::get(context, {}, returnTypes));
+  OwningOpRef<FuncOp> funcOp = cast<FuncOp>(Operation::create(state));
+
+  // create basic block and accompanying builder
+  Region &region = funcOp->getBody();
+  OpBuilder::InsertionGuard guard(builder);
+  auto *body = builder.createBlock(&region);
+  builder.setInsertionPointToStart(body);
+
+  // clear cache so that values are mapped to the right Basic Block
+  cache.clear();
+  for (auto init : inits) { toOp(init); }
+
+  // close with a fitting returnOp
+  std::vector<Value> testResults(states.size(), nullptr);
+  std::map<uint32_t, Value> undefOpsBySort;
+  uint32_t j = 0; // counters over inits vector
+  for (uint32_t i = 0, sz = states.size(); i < sz; ++i) {
+    if (states.at(i)->init > 0) {
+      // get the result of init's second argument since
+      // that is what we assign our state to  
+      testResults[i] = cache.at(inits.at(j++)->args[1]);
```

Comment:
the `inits.at(j++)` is what confuses me. `states.at(i)->init` is the line that defines the expression that
initializes sate `i`. What is the guarantee that it is the same expression as stored in `inits.at(j)`.

It is possible that `inits` is indexed by states. so `init.at(0)` is the init of `states.at(0)`, but then the else branch would not make sense since a state is skipped. 

It is possible that `inits` is just a vector of `init` commands in order they appear in the file. But then this is not necessarily the order of states. That is, state 2 can be initialized before state 1.

It is possible that `inits` is the list of all available initializations sorted by states. So if state 1 and 3 is initializes, then inits is `[init(1), init(3)]`. This is the only case in which the code makes sense, but that is a very ackward to explain design. I would rather have inits be `[init(1), null, init(3)]` and ensure that `inits` has same size as `states`. It would also remove the need to rely on `states.at(i)->init`, since init expression of `states.at(i)` would be `inits.at(i)`.

Reply:
Ah, I see what you mean! Will get that up in a jiffy :)